nifi-extension-bundles/nifi-aws-bundle/nifi-aws-kinesis/src/main/java/org/apache/nifi/processors/aws/kinesis/ConsumeKinesis.java

```diff
@@ -0,0 +1,737 @@
+        final RecordReaderFactory recordReaderFactory = context.getProperty(RECORD_READER).asControllerService(RecordReaderFactory.class);
+        final RecordSetWriterFactory recordWriterFactory = context.getProperty(RECORD_WRITER).asControllerService(RecordSetWriterFactory.class);
+        return new ReaderRecordProcessor(recordReaderFactory, recordWriterFactory, getLogger());
+    }
+
+    private static InitialPositionInStreamExtended getInitialPosition(final ProcessContext context) {
+        final InitialPosition initialPosition = context.getProperty(INITIAL_STREAM_POSITION).asAllowableValue(InitialPosition.class);
+        return switch (initialPosition) {
+            case TRIM_HORIZON ->
+                    InitialPositionInStreamExtended.newInitialPosition(InitialPositionInStream.TRIM_HORIZON);
+            case LATEST -> InitialPositionInStreamExtended.newInitialPosition(InitialPositionInStream.LATEST);
+            case AT_TIMESTAMP -> {
+                final String timestampValue = context.getProperty(STREAM_POSITION_TIMESTAMP).getValue();
+                final Instant timestamp = Instant.parse(timestampValue);
+                yield InitialPositionInStreamExtended.newInitialPositionAtTimestamp(Date.from(timestamp));
+            }
+        };
+    }
+
+    private static @Nullable MetricsFactory configureMetricsFactory(final ProcessContext context) {
+        final MetricsPublishing metricsPublishing = context.getProperty(METRICS_PUBLISHING).asAllowableValue(MetricsPublishing.class);
+        return switch (metricsPublishing) {
+            case DISABLED -> new NullMetricsFactory();
+            case LOGS -> new LogMetricsFactory();
+            case CLOUDWATCH -> null; // If no metrics factory was provided, CloudWatch metrics factory is used by default.
+        };
+    }
+
+    @OnStopped
+    public void onStopped() {
+        if (kinesisScheduler != null) {
+            shutdownScheduler();
+            kinesisScheduler = null;
+        }
+        if (kinesisClient != null) {
+            kinesisClient.close();
+            kinesisClient = null;
+        }
+        if (dynamoDbClient != null) {
+            dynamoDbClient.close();
+            dynamoDbClient = null;
+        }
+        if (cloudWatchClient != null) {
+            cloudWatchClient.close();
+            cloudWatchClient = null;
+        }
+
+        recordBuffer = null;
+        readerRecordProcessor = Optional.empty();
+    }
+
+    private void shutdownScheduler() {
+        final long start = System.nanoTime();
+        getLogger().debug("Shutting down Kinesis Scheduler");
+
+        boolean gracefulShutdownSucceeded;
+        try {
+            gracefulShutdownSucceeded = kinesisScheduler.startGracefulShutdown().get(KINESIS_SCHEDULER_GRACEFUL_SHUTDOWN_TIMEOUT.getSeconds(), SECONDS);
+            if (!gracefulShutdownSucceeded) {
+                getLogger().warn("Failed to shutdown Kinesis Scheduler gracefully. See the logs for more details");
```

Comment:
The `See the logs` portion is not necessary, recommend rewording:
```suggestion
                getLogger().warn("Failed to shutdown Kinesis Schedule after {} seconds", KINESIS_SCHEDULER_GRACEFUL_SHUTDOWN_TIMEOUT.getSeconds());
```

Reply:
"after {} seconds" part is true only when `TimeoutException` was thrown. 
Here we returned successfully, but for some reason didn't complete the shutdown successfully. I wouldn't mention the time here.

Since logs produced by KCL are not visible as bulletins, the failure reason won't be presented to a user via NiFi UI.
Too see the actual reason for graceful shutdown termination, application logs should be inspected.